4단계 - Interceptor 구현

testrace opened the pull request:
안녕하세요 용대님 😄 
용대님의 피드백 덕분에 많이 배우고 있습니다! 🔥 

이번 단계 요구사항은 비교적 단순했지만 학습 차원에서 스프링의 HandlerInterceptor를 기능을 따라 추가해 봤습니다.
미션 진행하면서 단위 테스트 작성하는 게 쉽지 않았는데요 mock 객체 등을 활용한 테스트가 미숙하다 보니
실제 객체나 fake 객체들을 많이 활용하는 편인데 그러다 보니 억지스러운 테스트를 한다는 생각도 들기도 합니다.

용대님은 단위 테스트 작성 시에 mock 객체를 활용하는 기준이 있으실까요? 
용대님의 의견 공유해 주시면 감사하겠습니다 🙏 

많은 피드백 부탁드립니다 🙇

testrace committed:
docs: 3단계 피드백 정리

testrace committed:
refactor: UserApiController 에 UserDao 적용

testrace committed:
test: RowMapperFunctionTest 에서 if 조건 추가

resultSet 의 내용을 확인 결과에 따라 분기처리되는 내용을 추가로 테스트할 수 있다

testrace committed:
refactor: RowMapperFunctionTest 에서 MockResultSet 적용

ResultSet 인터페이스를 일부 메서드를 구현하는 MockResultSet을 추가하여
Connection, Statement 를 사용하지 않고 ResultSet을 활용한 단위 테스트로 수정

testrace committed:
refactor: RequestParameterUtilsTest#re 메서드명 수정

re -> can_be_read_several_times
테스트의 의도를 가능한 명확하게 표현

testrace committed:
docs: 기능 목록 정의

testrace committed:
feat: Interceptor 인터페이스

preHandle: 컨트롤러 실행 전 수행
postHandle: 컨트롤러 실행 후 수행
afterCompletion: render 이후 수행
default 메서드를 제공하여 원하는 메서드만 오버라이드 할 수 있도록 한다

testrace committed:
feat: Add InterceptorRegistry

Interceptor 구현체를 추가할 수 있다

testrace committed:
feat: Add HandlerInterceptorRegistry#applyPreHandle

추가된 Interceptor 의 구현체들의 preHandle 메서드를 등록된 순서대로 실행한다.
실행 중 false 를 리턴 받으면 이후의 Interceptor 를 실행하지 않고 false 를 리턴한다.

testrace committed:
feat: Add HandlerInterceptorRegistry#applyPostHandle

추가된 Interceptor 의 구현체들의 postHandle 메서드를 등록된 순서대로 실행한다.

testrace committed:
feat: Add TimeTraceInterceptor

testrace committed:
feat: handlerExecutor#handler 전/후 HandleInterceptor 실행

testrace committed:
feat: View#render 이후 interceptor#applyAfterCompletion 수행

testrace committed:
feat: applyPreHandle 결과에 따라 진행 여부 결정

testrace committed:
test: stream allMatch study test

yongdae commented on src/test/java/core/jdbc/RowMapperFunctionTest.java:
```diff
@@ -112,4 +99,23 @@ void return_to_the_list_type() throws SQLException {
+        return createMockResultSet(null);
+    }
+
+    private static MockResultSet createMockResultSet(Map<String, Object> newRow) {
```
`용대님은 단위 테스트 작성 시에 mock 객체를 활용하는 기준이 있으실까요?` 
이런 의견을 주셨네요. :)

저는 대체적으로 기능을 테스트 할 때 `행위`까지 검증이 필요한 경우 mock을 활용하는 편입니다.

테스트를 함에 있어 활용하는 fake나 mock과 같은 구분(테스트 더블)을 고민하기보다는
테스트하는 기능의 어떤 부분까지 검증해야하는지 고민하는 편입니다. 

사실 이 부분은 크게 고민해본적이 없었던 부분인데 질문을 해주셔서 고민 해 볼 수 있는 계기가 되었네요.
감사합니다! 원하시는 의도의 답변이 아니라면 다시 의견 부탁드릴께요. :)

testrace replied:
충분한 답변이 되었습니다! 
의견을 공유해 주셔서 감사드립니다!! 😃 

테스트 코드를 접한 지 얼마 안 되어서 Input/Output이 명확하지 않으면
테스트 코드 작성할 때 '무엇(output)'으로 검증을 해야 하나 고민을 많이 하는 편이에요.
(그러다 보니 테스트를 위한 메서드를 만들고 싶은 생각이 자주 들더라고요!)

`행위`를 검증해야 하는 것에 대해서는 알듯 말듯 하지만, 덕분에 Input/Output 으로만 검증하기 보다 
기능에 따라 '어떤' 검증을 해야 할지에 대해서도 고민해 볼 수 있는 계기가 된 것 같습니다 🚀 

여담이지만 위 코드에서 MockResultSet이라고 이름을 지었지만 사실 FakeResultSet이 더 맞는 표현이 아닐까 생각도 드네요 😅

yongdae commented on src/test/java/core/jdbc/RowMapperFunctionTest.java:
```diff
@@ -2,99 +2,86 @@
-                rs.getString("name"),
-                rs.getString("email")
-            );
+        final MockResultSet mockResultSet = createMockResultSet();
+
+        final RowMapperFunction<Optional<User>> function = (rs) -> {
+            if (rs.next()) {
+                return Optional.of(new User(
+                    rs.getString("userId"),
+                    rs.getString("password"),
+                    rs.getString("name"),
+                    rs.getString("email")
+                ));
+            }
+            return Optional.empty();
```
검증하는 방식이 이전보다 간결해지고 더 명확해졌다고 생각해요. 💯

yongdae commented on README.md:
```diff
@@ -112,3 +112,41 @@
+> 각 Controller 메소드의 실행 속도를 측정한 후 debug level로 log를 출력하는 요구사항을 구현한다.  
+
+### 기능 목록
+- [x] Interceptor 인터페이스
+  - [x] 컨트롤러 실행 전 수행하는 메서드 (preHandle)
+  - [x] 컨트롤러 실행 후 수행하는 메서드 (postHandle)
+  - [x] 뷰 render 이후 수행하는 메서드 (afterCompletion)
+- [x] Interceptor 목록을 가지는 InterceptorRegistry 를 구현한다.
+  - [x] Interceptor 인터페이스의 구현체를 추가할 수 있는 메서드를 제공한다.
+  - [x] 추가된 Interceptor 의 구현체들의 preHandle 메서드를 등록된 순서대로 실행한다.
+    - [x] false 를 리턴 받으면 이후의 Interceptor 를 실행하지 않고 false 를 리턴한다.  
+  - [x] 추가된 Interceptor 의 구현체들의 postHandle 메서드를 등록된 순서대로 실행한다.
```
당연 할 수 있는 부분이지만 Interceptor를 꼼꼼하게 확인해서 기능 목록을 정의해주셨네요. 👍 
용주님이 가진 세심함이 엿보인다고 생각해요!

yongdae commented on src/main/java/core/mvc/DispatcherServlet.java:
```diff
@@ -52,9 +58,19 @@ protected void service(HttpServletRequest req, HttpServletResponse resp) throws
 
+            final Object handler = maybeHandler.get();
+            if (!handlerInterceptorRegistry.applyPreHandle(req, resp, handler)) {
+                resp.setStatus(HttpStatus.INTERNAL_SERVER_ERROR.value());
+                return;
+            }
+
+            ModelAndView mav = handlerExecutor.handle(req, resp, handler);
+
+            handlerInterceptorRegistry.applyPostHandle(req, resp, handler);
 
-            ModelAndView mav = handlerExecutor.handle(req, resp, maybeHandler.get());
             render(mav, req, resp);
+
+            handlerInterceptorRegistry.applyAfterCompletion(req, resp, handler);
```
미션에서 요구한 기능을 잘 구현해주셨네요. 👍

yongdae commented on src/main/java/core/mvc/DispatcherServlet.java:
```diff
@@ -38,6 +41,9 @@ public void init() {
 
         handlerExecutor = new HandlerExecutor(handlerAdapterRegistry);
+
+        handlerInterceptorRegistry = new HandlerInterceptorRegistry();
```
미션에서 요구한 사항은 아니지만 `addPathPatterns()` 와 같이
특정 패턴이나 url이 매칭되는 경우에 Interceptor가 동작 할 수 있게 만드는 방법에 대해서도
고민해보신다면 어떨까요? 🤔

물론 지금도 충분히 구현을 잘 해주셨다고 생각해요. 
이 의견의 반영 여부는 용주님이 선택해주시면 좋을 것 같습니다. :)

testrace replied:
앗 좋은 의견 감사합니다 👍 
평소엔 사용하면서도 이번 단계에서는 미처 생각을 못했네요.
지난 단계에서 배운 것을 복습 겸 구현해 보면 좋을 것 같아요! 😄

yongdae commented on src/main/java/core/mvc/DispatcherServlet.java:
```diff
@@ -52,9 +58,19 @@ protected void service(HttpServletRequest req, HttpServletResponse resp) throws
-            ModelAndView mav = handlerExecutor.handle(req, resp, maybeHandler.get());
             render(mav, req, resp);
+
+            handlerInterceptorRegistry.applyAfterCompletion(req, resp, handler);
```
applyAfterCompletion은 예외가 발생해도 수행 할 수 있게 만들어보면 어떨까요?

testrace replied:
applyAfterCompletion 메서드는 자주 사용하지 않아서 예외와 관련되어 있다는 걸 체크하지 못했네요 😢 
미션을 진행하면 할 수록 스프링 참 대단하다는 생각이 떠나질 않는 것 같아요 (용대님도요) 😃 
피드백 감사합니다!! 👍

yongdae commented on src/main/java/core/mvc/HandlerInterceptor.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface HandlerInterceptor {
+
+    default boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+        return true;
+    };
+
+    default void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+
+    };
+
+    default void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler) {
+
+    };
```
default method를 활용해주셨네요. 👍 
추가로 스프링의 [HandlerInterceptor](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/servlet/HandlerInterceptor.html) 예외를 정의해보면 어떨까요?

그리고 한 가지 질문이 있습니다!
용주님은 주로 어떤 경우에 default method를 사용하시나요? 🤔

testrace replied:
Interceptor에서 예외가 발생하는 경우를 고려하지 못했네요! 
Interceptor에서 예외를 던지게 하면 Interceptor를 사용하는 사용자에게 
적절한 예외처리를 미리 해야 한다는 의도까지 표현할 수도 있을 것 같아요 😃 

저는 개인적으로 인터페이스를 활용한 프로그래밍을 접한지가 얼마 안되어서 default method를 사용해본 경험이 없습니다.
이번 미션을 통해서 스프링5에서 HandlerInterceptor에 모두 default가 붙은 것으로 알고 있어서 따라 해봤습니다.
따라하면서 느낌 점은 공통 로직을 수행하거나, 인터페이스를 구현할 때 필요한 인터페이스만 구현할 수 있다는 장점이 있었습니다.
제가 인터페이스를 정의할 때 default method를 제공해야하는 경우가 생긴다면 위와 같은 이유로 사용할 것 같습니다! 😃 
(사실, 인터페이스에서 default method를 제공해야하는 경우가 얼마나 있을지는 잘 모르겠습니다 😅 )

용대님은 어떤 경우에 default method를 사용하시나요? 한 수 부탁드립니다 🙇

yongdae replied:
개인적으로는 default method를 활용해야하는
상황을 접한적이 별로 없어서 어떤 경우에 쓰는게 좋은지 명확히 답변드리기는 어렵네요. 😅 

다만 기존의 인터페이스를 확장해야 하는 경우에 사용했던 경험은 있습니다.
레거시한 프로젝트 환경에서 인터페이스를 확장해야하는 경우에 default method는 활용하기 좋은 수단이었습니다.

yongdae commented on src/main/java/core/mvc/HandlerInterceptorRegistry.java:
```diff
@@ -0,0 +1,36 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class HandlerInterceptorRegistry {
```
👍

yongdae commented on src/main/java/core/mvc/HandlerInterceptorRegistry.java:
```diff
@@ -0,0 +1,36 @@
+        interceptors.add(interceptor);
+    }
+
+    public boolean applyPreHandle(final HttpServletRequest request, final HttpServletResponse response, final Object handler) {
+        return interceptors.stream()
+            .allMatch(interceptor -> interceptor.preHandle(request, response, handler));
+    }
+
+    public void applyPostHandle(final HttpServletRequest request, final HttpServletResponse response, final Object handler) {
+        for (final HandlerInterceptor interceptor : interceptors) {
+            interceptor.postHandle(request, response, handler);
+        }
+    }
+
+    public void applyAfterCompletion(final HttpServletRequest request, final HttpServletResponse response, final Object handler) {
+        for (final HandlerInterceptor interceptor : interceptors) {
+            interceptor.afterCompletion(request, response, handler);
+        }
+    }
```
stream과 for문을 같이 사용하신 이유가 있으신가요?

testrace replied:
Collection의 데이터를 연산하거나 가공할 때는 stream을 사용하는 것이 가독성이 좋은 것 같고,
단순 반복하는 경우는 Iteratable.forEach 나 Stream.forEach 보다 for문이 가독성이 더 좋은 것 같아요 😃 
stream 사용이 익숙하지 못해서 그런 것 같기도 하네요 ㅎㅎ 😅 
예상밖의 질문인데요! 질문의 의도를 알려주시면 감사하겠습니다 🙏

yongdae replied:
음.. 이 질문을 드렸던 의도는
분명한 목적을 가지고 for문과 stream을 사용한다면 더 좋다고 생각해서
어떤 이유에서 for문과 stream을 같이 사용하셨는지 궁금해서였습니다!

Stream과 for문은 가독성의 차이도 있지만
Stream은 반복(Stream)과 처리를 나누어서 정의 할 수 있지만 for문은 대체적으로 2가지의 작업(반복문, 처리)을 하기 때문에
데이터들을 일련의 흐름을 통해 처리하는 경우에는 Stream을 활용하기를 추천드려요. :)

yongdae commented on src/main/java/core/mvc/tobe/interceptor/TimeTraceInterceptor.java:
```diff
@@ -0,0 +1,26 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TimeTraceInterceptor implements HandlerInterceptor {
```
시간을 측정하는 부분을 HandlerInterceptor를 잘 활용해서 만들어주셨네요. 👍

yongdae commented on src/main/java/core/mvc/tobe/interceptor/TimeTraceInterceptor.java:
```diff
@@ -0,0 +1,26 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(TimeTraceInterceptor.class);
+
+    private long startTime;
```
LocalDateTime을 활용한다면 소요시간 뿐만 아니라
필요에 따라 시작 시간과 종료 시간까지 활용 할 수 있을 것 같아요. :)

testrace replied:
시간 측정이라는 요구사항에 System#nanoTime 만 생각하고 있어서 
LocalDateTime으로 다양한 활용을 할 수 있을 거란 생각은 못했네요 👍 

생각해보니 Interceptor는 인스턴스가 1개이다 보니 인스턴스 변수를 사용하게 되면 동시성 이슈가 발생할 것 같아요.
Interceptor에서는 ThreadLocal을 적용하면 좋을 것 같은데 어떻게 생각하시나요?

yongdae replied:
이 부분까지 고민해주시다니 👍 
ThreadLocal을 사용하는게 좋다고 생각해요.

yongdae commented on src/test/java/core/mvc/HandlerInterceptorRegistryTest.java:
```diff
@@ -0,0 +1,237 @@
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+class HandlerInterceptorRegistryTest {
```
기능 목록 이상으로 mock까지 활용해서 테스트를 잘 만들어주셨네요. 💯

yongdae reviewed (changes requested):
피드백 드린 부분이 도움이 될 수 있어 다행이네요. 감사합니다. 🙇 
4단계 미션도 멋지게 구현해주셨네요. 👍 

여기서 마무리하기에 아쉬움이 남아 소소하지만 몇 가지 의견을 남겨놓았습니다.
확인하시고 다시 리뷰 요청 부탁드릴께요.

즐거운 금요일 저녁이 되셨으면 좋겠네요. 😄

testrace committed:
docs: 4단계 피드백 정리

testrace committed:
fix: TimeTraceInterceptor ThreadLocal 적용

System#nanoTime 을 적용하던 전역 변수를
thread-safe 동작을 위해 ThreadLocal 적용.
LocalDateTime을 적용하여 시작 시간과, 종료 시간, 소요 시간 모두
로그로 남길 수 있도록 변경.
예외 발생시에도 로그를 출력하고, ThreadLocal 도 안전하게 remove하게 위해
postHandle에서 수행하던 로직을 afterCompletion으로 이동.

testrace committed:
feat: Interceptor 의 메서드들에 `throws Exception` 추가

Interceptor 수행 중 예외가 발생할 수 있기 때문에
try-catch 로 직접 예외를 처리하거나
interceptor 를 수행하는 클라이언트에서 예외를 처리할 수 있도록 유도

testrace committed:
feat: applyAfterCompletion 는 예외가 발생하더라도 수행될 수 있도록 변경

개발자가 예외 등의 상황에서 후처리 가능하도록 예외 클래스를 인자로 추가

testrace committed:
feat: HandlerInterceptorExecution 추가

HandlerInterceptor 타입의 객체를 주입받아
HandlerInterceptorExecution 객체를 생성한다

testrace committed:
feat: HandlerInterceptorExecution#addPathPatterns

Interceptor가 수행 될 PathPattern을 여러개 등록할 수 있다

testrace committed:
feat: HandlerInterceptorExecution#getInterceptor(url)

url을 입력받아 PathPattern이 일치하는 경우 Intercepter를 반환한다.
PathPattern이 일치하지 않으면 아무것도 수행하지 않는 Intercepter를 반환한다.
등록된 PathPattern이 없는 경우 등록된 Interceptor를 반환한다.

testrace committed:
fix: DO_NOTHING_CHAIN_INTERCEPTOR 제거 및 Optional으로 변경

DO_NOTHING_CHAIN_INTERCEPTOR 의 가치가 없음으로 판단되어
Optional을 활용하도록 변경함

testrace committed:
feat: HandlerInterceptorRegistry2 추가

점진적 리팩토링을 위해 기존 Registry를 변경하지 않고
StranglerFig 패턴으로 리팩토링을 위해 어미에 2를 붙임
url 패턴별 interceptor들을 관리하는 HandlerInterceptorExecution의
일급 컬렉션으로 활용

testrace committed:
feat: Add UpdateUserAuthenticationInterceptor

회원정보 수정에만 적용되는 Interceptor 추가

testrace committed:
feat: Interceptor를 Registry에 등록할 때 url 패턴과 함께 등록

testrace committed:
refactor: Add HandlerInterceptorExecutor

InterceptorRegistry에서 수행하던 apply~ 메서드들을 HandlerInterceptorExecutor으로 이동

testrace committed:
refactor: HandlerInterceptorRegistry

StranglerFig Pattern 활용한 리팩토링

testrace committed:
refactor: HanderInterceptorExecutor 적용

testrace commented:
안녕하세요 용대님 😃 

피드백 주신 부분들도 반영을 하고, 추가로 제안해 주신 url 패턴에 따라 인터셉터가 동작할 수 있도록 구현해 봤습니다.
기존에는 인터셉터가 무조건 동작하도록 되어서 HandlerInterceptorRegistry 객체가 수행을 하면 되었는데,
url 패턴이 추가되면서 인터셉터 수행할 객체를 선정하는게 쉽지 않았던 것 같아요!

피드백 반영뿐만 아니라 5단계를 진행한 느낌이라 인터셉터에 대해서 조금 더 자세하게 공부할 수 있는 계기가 되어서 의미 있는 시간이었습니다!!
감사합니다 🙇

yongdae commented on src/main/java/core/mvc/DispatcherServlet.java:
```diff
@@ -38,25 +44,58 @@ public void init() {
+
+        handlerInterceptorRegistry = new HandlerInterceptorRegistry();
+        handlerInterceptorRegistry.addInterceptor(new TimeTraceInterceptor());
+        handlerInterceptorRegistry.addInterceptor(new UpdateUserAuthenticationInterceptor()).addPathPatterns("/api/users");
```
의견드린 부분을 멋지게 반영해주셨네요. 👍 

추가로 코드 길이도 길고, 인터셉터 등록과 패턴을 정의하고 있으니
줄바꿈을 통해서 2개의 작업을 하고 있다는 걸 알려주는 것도 좋을 것 같아요.

        handlerInterceptorRegistry.addInterceptor(new UpdateUserAuthenticationInterceptor())
                .addPathPatterns("/api/users");

yongdae commented on src/main/java/core/mvc/HandlerInterceptorExecution.java:
```diff
@@ -0,0 +1,65 @@
+
+    private boolean isUrlMatch(final String url) {
+        return pathPatterns.stream()
+            .anyMatch(pattern -> PathPatternUtil.isUrlMatch(pattern, url));
```
기존에 정의된 PathPatternUtil을 검증하고 활용해주셨네요. 💯

yongdae reviewed (approved):
의견드렸던 부분들을 적극적으로 수용해주셨네요. 💯 
특히 기존의 유틸리티도 활용해주신 코드가 좋았습니다!

그럼 이 미션은 여기서 마무리 짓도록 하겠습니다.
마지막까지 열심히 해주셔서 재미있게 미션을 함께 할 수 있었습니다. 🙇 

수고하셨습니다! 당연히 미션을 완주하시겠지만 끝까지 힘내주세요. 🔥🔥
